# Data Science Academy

## Fundamentals of Python Language for Data Analyse and Data Science

### Manipulation of Data with Pandas

In [1]:
# Python Language version
from platform import python_version
print("Python Language version used in this Jupyter Notebook: ", python_version())

Python Language version used in this Jupyter Notebook:  3.11.5


### User Guide Pandas:

https://pandas.pydata.org/docs/user_guide/index.html

In [2]:
# Install the exact package version (if necessary)
!pip install -q pandas==1.5.3

In [3]:
import pandas as pd

In [4]:
pd.__version__

'2.0.3'

### Manipulating Data in DataFrames from Pandas

In [5]:
# Create a dictionary
data = {"State": ["Santa Catarina", "Rio de Janeiro", "Pernambuco", "Bahia", "Minas Gerais"], 
        "Year": [2004, 2005, 2006, 2007, 2008],
        "Employement Rate": [1.5, 1.7, 1.6, 2.4, 2.7]}

In [6]:
# Import the function DataFrame from Pandas
from pandas import DataFrame

In [7]:
# Convert the dicitonary in a dataframe
df = DataFrame(data)

In [8]:
# Visualize the first 5 lines
df.head()

State  Year  Employement Rate
0  Santa Catarina  2004               1.5
1  Rio de Janeiro  2005               1.7
2      Pernambuco  2006               1.6
3           Bahia  2007               2.4
4    Minas Gerais  2008               2.7

In [9]:
type(df)

pandas.core.frame.DataFrame

In [10]:
# Reorganizing the columns
DataFrame(data, columns = ["State", "Employement Rate", "Year"])

State  Employement Rate  Year
0  Santa Catarina               1.5  2004
1  Rio de Janeiro               1.7  2005
2      Pernambuco               1.6  2006
3           Bahia               2.4  2007
4    Minas Gerais               2.7  2008

In [11]:
df2 = DataFrame(data, columns=["State", "Employement Rate", "Grow Rate", "Year",],
               index = ["state1", "state2", "state3", "state4", "state5"])

In [12]:
df2

State  Employement Rate Grow Rate  Year
state1  Santa Catarina               1.5       NaN  2004
state2  Rio de Janeiro               1.7       NaN  2005
state3      Pernambuco               1.6       NaN  2006
state4           Bahia               2.4       NaN  2007
state5    Minas Gerais               2.7       NaN  2008

In [13]:
df2.columns

Index(['State', 'Employement Rate', 'Grow Rate', 'Year'], dtype='object')

In [14]:
# Printing only one column from the DataFrame
df2["State"]

state1    Santa Catarina
state2    Rio de Janeiro
state3        Pernambuco
state4             Bahia
state5      Minas Gerais
Name: State, dtype: object

In [15]:
# Printing only two columns from the DataFrame
df2[["Employement Rate", "Year"]]

Employement Rate  Year
state1               1.5  2004
state2               1.7  2005
state3               1.6  2006
state4               2.4  2007
state5               2.7  2008

In [16]:
df2.index

Index(['state1', 'state2', 'state3', 'state4', 'state5'], dtype='object')

In [17]:
df2.values

array([['Santa Catarina', 1.5, nan, 2004],
       ['Rio de Janeiro', 1.7, nan, 2005],
       ['Pernambuco', 1.6, nan, 2006],
       ['Bahia', 2.4, nan, 2007],
       ['Minas Gerais', 2.7, nan, 2008]], dtype=object)

In [18]:
df2.dtypes

State                object
Employement Rate    float64
Grow Rate            object
Year                  int64
dtype: object

In [19]:
# Python Language is case sensitive
# df2["state"]

In [20]:
df2.filter(items = ["state3"], axis = 0)

State  Employement Rate Grow Rate  Year
state3  Pernambuco               1.6       NaN  2006

### Using NumPy and Pandas for Data Manipulation

In [21]:
# Importing Numpy
import numpy as np

In [22]:
df2.head()

State  Employement Rate Grow Rate  Year
state1  Santa Catarina               1.5       NaN  2004
state2  Rio de Janeiro               1.7       NaN  2005
state3      Pernambuco               1.6       NaN  2006
state4           Bahia               2.4       NaN  2007
state5    Minas Gerais               2.7       NaN  2008

In [23]:
df2.describe()

Employement Rate         Year
count          5.000000     5.000000
mean           1.980000  2006.000000
std            0.535724     1.581139
min            1.500000  2004.000000
25%            1.600000  2005.000000
50%            1.700000  2006.000000
75%            2.400000  2007.000000
max            2.700000  2008.000000

In [24]:
df2.isna()

State  Employement Rate  Grow Rate   Year
state1  False             False       True  False
state2  False             False       True  False
state3  False             False       True  False
state4  False             False       True  False
state5  False             False       True  False

In [25]:
df2["Grow Rate"].isna()

state1    True
state2    True
state3    True
state4    True
state5    True
Name: Grow Rate, dtype: bool

In [26]:
df2["Grow Rate"] = np.arange(5.)

In [27]:
df2

State  Employement Rate  Grow Rate  Year
state1  Santa Catarina               1.5        0.0  2004
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006
state4           Bahia               2.4        3.0  2007
state5    Minas Gerais               2.7        4.0  2008

In [28]:
df2.dtypes

State                object
Employement Rate    float64
Grow Rate           float64
Year                  int64
dtype: object

In [29]:
df2["Grow Rate"].isna()

state1    False
state2    False
state3    False
state4    False
state5    False
Name: Grow Rate, dtype: bool

In [30]:
# Statistic resume of Dataframe
df2.describe()

Employement Rate  Grow Rate         Year
count          5.000000   5.000000     5.000000
mean           1.980000   2.000000  2006.000000
std            0.535724   1.581139     1.581139
min            1.500000   0.000000  2004.000000
25%            1.600000   1.000000  2005.000000
50%            1.700000   2.000000  2006.000000
75%            2.400000   3.000000  2007.000000
max            2.700000   4.000000  2008.000000

### Slicing Pandas Dataframes

In [31]:
df2

State  Employement Rate  Grow Rate  Year
state1  Santa Catarina               1.5        0.0  2004
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006
state4           Bahia               2.4        3.0  2007
state5    Minas Gerais               2.7        4.0  2008

In [32]:
df2["state2": "state4"]

State  Employement Rate  Grow Rate  Year
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006
state4           Bahia               2.4        3.0  2007

In [33]:
df2[df2["Employement Rate"] < 2]

State  Employement Rate  Grow Rate  Year
state1  Santa Catarina               1.5        0.0  2004
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006

In [34]:
df2[["State", "Grow Rate"]]

State  Grow Rate
state1  Santa Catarina        0.0
state2  Rio de Janeiro        1.0
state3      Pernambuco        2.0
state4           Bahia        3.0
state5    Minas Gerais        4.0

In [35]:
df2[["State", "Grow Rate", "Employement Rate"]]

State  Grow Rate  Employement Rate
state1  Santa Catarina        0.0               1.5
state2  Rio de Janeiro        1.0               1.7
state3      Pernambuco        2.0               1.6
state4           Bahia        3.0               2.4
state5    Minas Gerais        4.0               2.7

### Filling Absent Values in Pandas DataFrames

The function fillna() is used to fill absent values. The function offers many options. We can use a specific value, an aggregate functon (for instance, average() or the previous values or the next.

For that example we'll use the mode, the statistic represents the value that appear more times in a variable

In [36]:
# First we import a dataset
df = pd.read_csv("dataset.csv")

In [37]:
df.head(5) # 5 is the default value

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [38]:
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [39]:
# We extract the mode of Quantity column
mode = df["Quantidade"].value_counts().index[0]

In mathematics, the mode is the value that appears most frequently in a set of data. It's a measure of central tendency, like the mean and median, used to describe the typical value in a dataset. 

In data analysis, the mode helps identify the most common occurrence or trend within a dataset, which can be useful for various applications. 

In [40]:
print(mode)

3.0


In [41]:
# And in the end we fill the values NA with the mode
df["Quantidade"].fillna(value = mode, inplace = True) 
# inplace = True, it tells pandas to replace the values in the dataframe
# itself instead of creating a copy

In [42]:
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

### Data Query in Pandas DataFrame

With the pandas, we can create data frames. which are essentially tables. In such a way, we can make queries. And for that, we use the consult method query(). See the example below:

In [43]:
# We check the minimum and max values of the colum Valor_Venda
df.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

The sales interval (Valor_Venda) is of 0.44 to 22638. Let's make a consultation and return all the sales inside a range of values. We make this as the instruction below:

In [44]:
# We generate a new dataframe only with the sales interval 
# between 229 and 10000
df2 = df.query("229 < Valor_Venda < 10000")

In [45]:
# Then, we confirm the minimum and max values
df2.Valor_Venda.describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [46]:
# We generate a new dataframe only with the sale values above the average
df3 = df2.query("Valor_Venda > 766")

In [47]:
df3.head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
3   US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
7   CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
10  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
11  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
24  CA-2015-106320  2015-09-25   EB-13870  Consumer  United States   West   

         ID_Produto   Categoria  \
3   FUR-TA-10000577   Furniture   
7   TEC-PH-10002275  Technology   
10  FUR-TA-10001539   Furniture   
11  TEC-PH-10002033  Technology   
24  FUR-TA-10000577   Furniture   

                                     Nome_Produto  Valor_Venda  Quantidade  
3   Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
7                  Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
10       Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11  Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
24  Bretford CR4500 Series Slim Rectangular Table    1044.6300         3.0

Query executed succesfully!!

### Checking the Occurrence of Several Values in a Column

In our example dataset we have the column Quantidade that represents the quantity of sold items in each one of the sales. Imagine that we need to know in which sales were sold 5, 7, 9 or 11 items.

Will we apply that type of filter in our dataframe?

Easy. Pandas offers the method isin() to check several values in a column. Who knows SQL Language must already have noticed that the method is the same as the clause IN in SQL. Let's go to the example.

In [48]:
df.shape

(9994, 11)

In [49]:
# Then we apply the filter
df[df["Quantidade"].isin([5, 7, 9, 11])]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
...              ...         ...        ...          ...            ...   
9974  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9977  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9982  US-2016-157728  2016-09-22   RC-19960     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
5        West  FUR-FU-10001487        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
...       ...              ...              ...   
9974     West  OFF-AR-10004752  Office Supplies   
9977     West  OFF-FA-10003467  Office Supplies   
9981     East  OFF-LA-10004484  Office Supplies   
9982  Central  OFF-PA-10002195  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
...                                                 ...          ...   
9974                               Blackstonian Pencils      18.6900   
9977           Alliance Big Bands Rubber Bands, 12/Pack      13.8600   
9981                                          Avery 476      16.5200   
9982  RSVP Cards & Envelopes, Blank White, 8-1/2" X ...      35.5600   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
5            7.0  
9            5.0  
10           9.0  
14           5.0  
...          ...  
9974         7.0  
9977         7.0  
9981         5.0  
9982         7.0  
9988         5.0  

[2128 rows x 11 columns]

In the instruction above, we're filtering the dataframe called f, returning all the lines where the column Quantity is equal to the values 5, 7, 9 or 11. We give a list of values as argument for the method isin().

Let's make it a little more fun. If you executed the instruction above, you noticed that were returned 2.128 lines. And if we want to return only 10 lines? Just slice the result like this:

In [50]:
# Shape
df[df["Quantidade"].isin([5, 7, 9, 11])].shape

(2128, 11)

In [51]:
df[df["Quantidade"].isin([5, 7, 9, 11])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
5      West  FUR-FU-10001487        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
36  Central  FUR-FU-10003664        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
21                                         Newell 318      19.4600         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length      60.3400         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...    3083.4300         7.0  
35                                        GE 30524EE4    1097.5440         7.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0

Isn't Pandas incredible?

### Logical Operators for Data Manipulation with Pandas

The Logical Operators are excelent to filter dataframes and return exactly the data we need for our work. To know more about the logical operators rules, go to this page:

https://press.rebus.community/programmingfundamentals/chapter/logical-operators/

FIrst we'll use the logical operator AND to check two conditions. It'll be returned the registers when both of the conditions were true

In [55]:
# FIltering the sales that happened for working from 
# home segment and in the South region
df[ (df.Segmento == "Home Office") & (df.Regiao == "South") ].head()

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais Regiao  \
182  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
183  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
184  CA-2014-158274  2014-11-19   RM-19675  Home Office  United States  South   
231  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   
232  US-2017-100930  2017-04-07   CS-12400  Home Office  United States  South   

          ID_Produto   Categoria  \
182  TEC-PH-10003273  Technology   
183  TEC-PH-10004896  Technology   
184  TEC-AC-10002345  Technology   
231  FUR-TA-10001705   Furniture   
232  FUR-TA-10003473   Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
182                                      AT&T TR1909W     503.9600         4.0  
183                        Nokia Lumia 521 (T-Mobile)     149.9500         5.0  
184                 HP Standard 104 key PS/2 Keyboard      29.0000         2.0  
231  Bush Advantage Collection Round Conference Table     233.8600         2.0  
232        Bretford Rectangular Conference Table Tops     620.6145         3.0

But it might be needed to check two conditions and return the register if at least one of them are true. In that case, we use the operator OR, like we show bellow

In [54]:
# Filtering the sales that happened for working from 
# home or South region
df[(df.Segmento == "Home Office") | (df.Regiao == "South")].tail()

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
9979  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9980  US-2015-151435  2015-09-06   SW-20455     Consumer  United States   
9987  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   
9989  CA-2014-110422  2014-01-21   TB-21400     Consumer  United States   

     Regiao       ID_Produto        Categoria  \
9979   West  OFF-BI-10002026  Office Supplies   
9980  South  FUR-TA-10001029        Furniture   
9987  South  TEC-AC-10001539       Technology   
9988  South  TEC-PH-10004006       Technology   
9989  South  FUR-FU-10001889        Furniture   

                                           Nome_Produto  Valor_Venda  \
9979                  Ibico Recycled Linen-Style Covers      437.472   
9980                         KI Adjustable-Height Table       85.980   
9987  Logitech G430 Surround Sound Gaming Headset wi...       79.990   
9988                    Panasonic KX - TS880B Telephone      206.100   
9989                             Ultra Door Pull Handle       25.248   

      Quantidade  
9979        14.0  
9980         1.0  
9987         1.0  
9988         5.0  
9989         3.0

The not operator is the opposite of the first example.

In [56]:
# Filtering the sales that neither happened for working from 
# home or South region
df[(df.Segmento != "Home Office") & (df.Regiao != "South")].sample(5)

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais   Regiao  \
668   CA-2014-156314  2014-12-24   RP-19390  Consumer  United States     East   
2703  CA-2014-143413  2014-05-24   RP-19390  Consumer  United States     East   
9844  CA-2014-163867  2014-06-03   RE-19450  Consumer  United States  Central   
5227  CA-2016-119865  2016-06-21   AS-10090  Consumer  United States     East   
2366  CA-2014-102295  2014-11-24   EH-13990  Consumer  United States     West   

           ID_Produto        Categoria  \
668   FUR-FU-10003096        Furniture   
2703  OFF-PA-10002319  Office Supplies   
9844  FUR-FU-10001475        Furniture   
5227  FUR-BO-10003272        Furniture   
2366  FUR-CH-10001714        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
668        Master Giant Foot Doorstop, Safety Yellow       30.360         5.0  
2703                                      Xerox 1944      116.280         3.0  
9844                      Contract Clock, 14", Brown       61.544         7.0  
5227  O'Sullivan Living Dimensions 5-Shelf Bookcases      353.568         2.0  
2366  Global Leather & Oak Executive Chair, Burgundy      120.712         1.0

### Grouping Data in DataFrames with Group By

The Pandas function Groupby is a versatile and easy-to-use function that helps to take a general vision of data. That makes it easier to explore the dataset and reveal the relationship between the variables.

The following code will group the lines based in combinations Segmento/Regiao/Valor_Venda, and it will give the sales average rate of each group.

In [57]:
# Applying group by
df[["Segmento", "Regiao", "Valor_Venda"]].groupby(["Segmento", "Regiao"]).mean()

Valor_Venda
Segmento    Regiao              
Consumer    Central   207.946728
            East      238.875539
            South     233.390180
            West      217.033955
Corporate   Central   234.763466
            East      228.516929
            South     238.992025
            West      235.265911
Home Office Central   208.248046
            East      253.911805
            South     272.996329
            West      239.442692

In the above instruction we filter the data extracting 3 columns: ['Segmento', 'Region', 'Valor_Venda']. In the sequence, we group for two columns: ['Segmento', 'Regiao]. And then we calculate the average for the column that were from group by, in that case the column Sales.

The behavior of group by with Pandas is the same observed in SQL Language

### Multiple Agregation with Group By

Let's explore more the groupby() function  because we have several options of simple data summarization. In the example below, we'll unify the function groupby() with the function agg() that realizes multiple aggregation

In [59]:
# Apply group by
df[["Segmento", "Regiao", "Valor_Venda"]].groupby(["Segmento", "Regiao"]).agg("mean", "std", "count")

Valor_Venda
Segmento    Regiao              
Consumer    Central   207.946728
            East      238.875539
            South     233.390180
            West      217.033955
Corporate   Central   234.763466
            East      228.516929
            South     238.992025
            West      235.265911
Home Office Central   208.248046
            East      253.911805
            South     272.996329
            West      239.442692

In the instruction above, first we filter the data extracting 3 columns: ['Segmento', 'Region', 'Valor_Venda']. Following, we group by two columns: ['Segmento', 'Region']. And, hen we aggreagate the data calculating the average, standard deviation and the count of elements for the column that was out of the group by, in this case the column Valor_Venda.

The function agg() receives as argument a list of functions for aggregation.

### Filtering Pandas DataFrame Based in Strings

Pandas offers several functions for strings manipulations. We'll start with strings filters based in initial and final letters.


In [60]:
df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [61]:
# We filter the dataframe by Segmento column with values 
# that starts with the letters "Con"
df[df.Segmento.str.startswith("Con")].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0

In [65]:
# We filter the dataframe by Segmento column with values 
# that starts with the letters "Con"
df.Segmento.value_counts()

Segmento
Consumer       5191
Corporate      3020
Home Office    1783
Name: count, dtype: int64

In [62]:
# We filter the dataframe by Segmento column with values 
# that ends with the letters "mer"
df[df.Segmento.str.endswith("mer")].head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520  Consumer  United States  South   
3  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
5  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   
5  FUR-FU-10001487        Furniture   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0  
5  Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0

The functions startswith() and endswith() are very useful when it's necessary to filter strings by character that appears in the beginning and/or ending.

### Split the strings in Pandas DataFrames

With Pandas we can realize several split tasks spliting a column or extracting elements of our interest. Let's see the example.

In [66]:
df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         3.0  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [67]:
df["ID_Pedido"].head()

0    CA-2016-152156
1    CA-2016-152156
2    CA-2016-138688
3    US-2015-108966
4    US-2015-108966
Name: ID_Pedido, dtype: object

This is the data format of the column "ID_Pedido"

* CA-2016-152156
* US-2015-108966

We have country, the year and the order id. Let's split that colum and extract the year to record it in a new column:

In [68]:
# Split the column by character '-'
df["ID_Pedido"].str.split("-")

0       [CA, 2016, 152156]
1       [CA, 2016, 152156]
2       [CA, 2016, 138688]
3       [US, 2015, 108966]
4       [US, 2015, 108966]
               ...        
9989    [CA, 2014, 110422]
9990    [CA, 2017, 121258]
9991    [CA, 2017, 121258]
9992    [CA, 2017, 121258]
9993    [CA, 2017, 119914]
Name: ID_Pedido, Length: 9994, dtype: object

- Note that the results are lists in Python. To extract the year, we need to specify the index of the position we want to extract (in our case, position 2, so index 1 in Python):

In [69]:
df["ID_Pedido"].str.split('-').str[1].head()

0    2016
1    2016
2    2016
3    2015
4    2015
Name: ID_Pedido, dtype: object

In [70]:
# We make the column split and extract the item in the second position (index 1)
df["Year"] = df["ID_Pedido"].str.split('-').str[1]

In [71]:
# Then we check the new column created
df.head()

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase     261.9600         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         3.0   
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0   
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0   
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0   

   Year  
0  2016  
1  2016  
2  2016  
3  2015  
4  2015

### Strip the Strings in Pandas Dataframes

Be careful not to confuse. We saw Split and now we'll see Strip. They are different functions.

In [73]:
df.head(3)

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  \
0                  Bush Somerset Collection Bookcase       261.96         3.0   
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...       731.94         3.0   
2  Self-Adhesive Address Labels for Typewriters b...        14.62         2.0   

   Year  
0  2016  
1  2016  
2  2016

In [76]:
df["Data_Pedido"].head(3)

0    2016-11-08
1    2016-11-08
2    2016-06-12
Name: Data_Pedido, dtype: object

The column 'Data_Pedido' is the product ship date in the format YYYY-MM-DD. Imagine that is necessary to let only with two digits without altering the type of the variable. We make that with the function lstrip(), which means, left strip.

In [75]:
# Let's remove the digits 2 and 0 from the left of the variable value 'Data_Pedido'
df['Data_Pedido'].str.lstrip('20')

0       16-11-08
1       16-11-08
2       16-06-12
3       15-10-11
4       15-10-11
          ...   
9989    14-01-21
9990    17-02-26
9991    17-02-26
9992    17-02-26
9993    17-05-04
Name: Data_Pedido, Length: 9994, dtype: object

As we don't use inplace = True the modifications is only in the memory and doesn't alter the dataframe. We can still use the functions rstrip() and strip() with different vairations of strings strip.


### Strings Replace in Pandas Dataframes

If it's necessary to replace characters in a string Pandas offers a function for that too. For instance, let's replace 2 characters in one of the columns.

## End
### Thank you!